Feature/gamestatemachine

BNS-MarkUlrich opened the pull request.

BNS-MarkUlrich committed:
WIP Adding Generic GameStateMachine

BNS-MarkUlrich committed:
Merge branch 'develop' into feature/gamestatemachine

BNS-MarkUlrich committed:
Merge branch 'develop' into feature/gamestatemachine

BNS-MarkUlrich committed:
Removed unused code, added singleton isntance

bas-boop commented on Assets/Scripts/Utils/SingletonInstance.cs:
```diff
@@ -0,0 +1,42 @@
+    {
+        transform.parent = transform.root;
+
+        if (Instance != this)
+            Destroy(gameObject);
```
Would be nice if there was an error message here. Some along the lines of "There are multiple of the same singletons."

BNS-MarkUlrich replied:
I don't think this is necessary as it automatically fixes the problem before the new instance is ever able to even call any code as it is in OnEnable.

bas-boop commented on Assets/Scripts/GameStateMachine/Example/ExampleGameState.cs:
```diff
@@ -0,0 +1,18 @@
+        base.EnterState();
+        SetNextState<ExampleBootState>();
+
+        LoadScene("StatemachineTest");
```
Why is this string hardcoded? In ExampleBootState.cs do you it in a better way.

bas-boop commented on Assets/Scripts/GameStateMachine/GameStateMachine.cs:
```diff
@@ -0,0 +1,80 @@
+    /// <param name="state">The State subscribing itself to the StateMachine.</param>
+    public static void Subscribe(this State state)
+    {
+        for (int i = 0; i < States.Count; i++)
+            if (States[i].ToString() == state.ToString())
+                return;
```
```suggestion
        for (int i = 0; i < States.Count; i++)
        {
            if (States[i].ToString() == state.ToString())
                return;
        }
```
I like it when something without a scope that does technically has another scope in it to add a scope then.

BNS-MarkUlrich replied:
This better?
```csharp
    public static void Subscribe(this State state)
    {
        if (!States.Any(s => s.ToString() == state.ToString()))
            States.Add(state);
    }
   ```

bas-boop commented on Assets/Scripts/GameStateMachine/GameStateMachineBehaviour.cs:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    private void DebugStateInstance() => print("I work!");
+    #endregion
```
Is this still needed? If not it can be removed. Or maybe the class name should be an example.

bas-boop commented on Assets/Scripts/GameStateMachine/State.cs:
```diff
@@ -0,0 +1,76 @@
+        OwningGameStateMachine = GameStateMachine.OwningGameStateMachine;
+    }
+    
+    protected void SetNextState<TState>() where TState: State, new() 
+        => _nextState = GameStateMachine.GetState<TState>();
```
I'm just wondering how the syntax here works. If you can explain that would be great.

BNS-MarkUlrich replied:
The method is generic, as indicated by `<TState>`, which means it can work with any type that is or inherits from `State`. The `where TState: State, new()` LINQ expression constraint ensures that `TState` is or derives from `State` and has a parameterless constructor. This is necessary because the method needs to create an instance of `TState`.

bas-boop reviewed (commented):
Nice code has been made.

I do have some common feedback:

1. Classes summary
2. Lake of namespaces
3. Lake of Tooltips/summary with some complex variables (GameStateMachine.cs)

BNS-MarkUlrich committed:
Implemented Feedback

bas-boop commented on Assets/Scripts/GameStateMachine/GameStateMachine.cs:
```diff
@@ -0,0 +1,94 @@
+        /// <summary>
+        /// Moves the current state to the next state set in the state.
+        /// </summary>
+        internal static void MoveToNextState() => CurrentState.MoveToNextState();
```
```suggestion
        internal static void MoveToNextState()
            => CurrentState.MoveToNextState();
```
Consistant with all the other functions.

bas-boop reviewed (approved):
1 little thing, but we can look through the fingers.

BNS-MarkUlrich committed:
Implemented accessibility features to StateMachine

BNS-MarkUlrich committed:
Updated SingletonInstance to use namespace

BNS-MarkUlrich committed:
Fixed Statemachine not using correct namespace

BNS-MarkUlrich committed:
Added debug mode toggle, and:

- Fixed namespaces for states

BNS-MarkUlrich committed:
Code Cleanup, and:

- Added TODO's to State class related to child base functions
- Moved LoadScene method to bottom
- Renamed StateMachineHashcode to just Hashcode
- Fixed ExampleGameState moving back to BootSstate instead of MainMenuState

BNS-MarkUlrich committed:
Moved files into proper folder structure, and:

- OwningStateMachine in States now refers to actual Statemachine from owner instance, making functions in StateMachineInstance obsolete
- Added ExampleStateMachineInstance for storing of sample code related to StateMachineInstance

BNS-MarkUlrich committed:
Removed unused methods in StateMachineInstance, and:

- Updated summaries in ExampleStateMachineInstance;
- Moved sample states to States folder inside Samples folder;

BNS-MarkUlrich committed:
Fixed deprecated function in SingletonInstance

BNS-MarkUlrich committed:
Cleaned up naming conventions, and:

- Added Todos to StateMachineInstance and base State
- Added colouring to debug messages from StateMachine
- Changed States in StateMachine from List to HashSet for performance
- Changed accesibility of Name variable in State from private to public
- Renamed CurrentStateIs methods in StateMachine to IsCurrentState for better readability

BNS-MarkUlrich committed:
Added abstract functions for states, and:

- Expanded OnStateEnter and Exit trigger showcase in ExampleStateMachineInstance

BNS-MarkUlrich committed:
Added Instance property to StateMachine, and:

- Renamed StateMachineInstance to StateMachineBehaviour to better reflect its use
- Refactored SetState method in StateMachine to also return the State it set
